File: 12.Manacher'sAlgo.py
# Manacher’s Algorithm is an efficient algorithm to find the longest palindromic substring in a given string.
# It runs in O(n) time, making it much faster than the naive O(n²) or dynamic programming approaches.

# Explanation

# A palindrome is a string that reads the same forwards and backwards (like racecar, aba).

# The brute-force approach checks every possible substring → O(n²).

# Manacher’s Algorithm improves this by using previously computed palindrome lengths and symmetry to skip redundant checks.

# It works by:

# Transforming the string with separators (#) so even and odd length palindromes can be handled uniformly.

# Example: "abba" → ^#a#b#b#a#$ (^ and $ are sentinels).

# Maintaining an array P where P[i] stores the radius of palindrome centered at position i.

# Using two variables:

# C → the center of the current palindrome.

# R → right boundary of the current palindrome.

# Expanding palindromes while reusing results from mirror indices.

# Result: The maximum value in P gives the length and position of the longest palindrome substring.

def manacher(s):
    t = "^#" + "#".join(s) + "#$"
    n = len(t)

    P = [0] * n
    C, R = 0, 0

    for i in range(1, n - 1):
        mirror = 2*C - i

        if i < R:
            P[i] = min(R - i, P[mirror])

        while t[i + 1 + P[i]] == t[i - 1 - P[i]]:
            P[i] += 1

        if i + P[i] > R:
            C, R = i, i + P[i]

    max_len = max(P)
    center_index = P.index(max_len)

    start = (center_index - max_len) // 2
    return s[start: start + max_len]

s = "babad"
print("Longest Palindromic Substring:", manacher(s))






  

File: 12.Manacher'sAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )                                          */
/*  ___ _                      _              _                                  */
/*  | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)                                */
/*  | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |                                */
/*  |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |                                */
/*                                     |__/                                      */
/* ----------------------------------------------------------------------------  */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                         */
/*    Github : https://github.com/Manu577228                                    */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio         */
/* ----------------------------------------------------------------------------  */

import java.io.*;

public class ManacherAlgo {

    // Function implementing Manacher's Algorithm
    public static String manacher(String s) {
        // Step 1: Transform the string by inserting '#' between characters
        StringBuilder sb = new StringBuilder("^#");
        for (int i = 0; i < s.length(); i++) {
            sb.append(s.charAt(i)).append("#");
        }
        sb.append("$");
        String t = sb.toString(); // Example: "babad" -> "^#b#a#b#a#d#$"

        int n = t.length();
        int[] P = new int[n];   // Step 2: Array to store palindrome radius
        int C = 0, R = 0;       // Center (C) and right boundary (R)

        // Step 3: Loop through transformed string
        for (int i = 1; i < n - 1; i++) {
            int mirror = 2 * C - i; // Mirror index of i around center C

            // If within current palindrome boundary
            if (i < R) {
                P[i] = Math.min(R - i, P[mirror]);
            }

            // Try to expand palindrome around i
            while (t.charAt(i + 1 + P[i]) == t.charAt(i - 1 - P[i])) {
                P[i]++;
            }

            // If palindrome expands beyond R, update C and R
            if (i + P[i] > R) {
                C = i;
                R = i + P[i];
            }
        }

        // Step 4: Find max length palindrome
        int maxLen = 0;
        int centerIndex = 0;
        for (int i = 1; i < n - 1; i++) {
            if (P[i] > maxLen) {
                maxLen = P[i];
                centerIndex = i;
            }
        }

        // Step 5: Extract substring from original string
        int start = (centerIndex - maxLen) / 2;
        return s.substring(start, start + maxLen);
    }

    // Driver code
    public static void main(String[] args) throws IOException {
        String s = "babad";
        String result = manacher(s);
        System.out.println("Longest Palindromic Substring: " + result);
    }
}
